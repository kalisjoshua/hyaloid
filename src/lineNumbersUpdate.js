import range from './range';

const getLineNumbers = (el) => el.lineNumbers.querySelectorAll('ins');

function lineNumbersUpdate(element, insert) {
  const lineNumberElements = getLineNumbers(element);

  /* istanbul ignore else */
  if (element.lineCount !== lineNumberElements.length) {
    /* istanbul ignore else */
    if (element.lineCount > lineNumberElements.length) {
      const diff = element.lineCount - lineNumberElements.length;

      range(diff).forEach(insert);
    } else if (element.lineCount < lineNumberElements.length) {
      const diff = lineNumberElements.length - element.lineCount;
      const numbers = Array.from(lineNumberElements);

      numbers
        .slice(numbers.length - diff)
        .forEach((el) => el.remove());
    }
  }

  /* eslint no-param-reassign:0 */
  // This block of code specifically needs to edit the DOM
  // and therefor break the rules of ESLint temporarily.
  Array.from(getLineNumbers(element))
    .forEach((line, index) => {
      if (element.lineError[index + 1]) {
        line.dataset.hasError = true;
        line.setAttribute('title', element.lineError[index + 1]);
      } else {
        line.dataset.hasError = null;
        line.removeAttribute('title');
      }
    });

  element.editor.style.marginLeft = `${element.lineNumbers.offsetWidth}px`;
  element.editor.style.width = `calc(100% - ${element.editor.style.marginLeft})`;
  /* eslint no-param-reassign:1 */

  setTimeout(() => lineNumbersUpdate(element, insert), 100);
}

export default lineNumbersUpdate;
