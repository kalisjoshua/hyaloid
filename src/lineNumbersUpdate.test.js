import lineNumbersUpdate from './lineNumbersUpdate';

jest.useFakeTimers();

let lineNumbers = [];
let mockEditor;

function mockLineNumberFactory() {
  const obj = Object.create({
    remove() {
      lineNumbers = lineNumbers.slice(1);
    },
    removeAttribute(key) {
      delete obj[key];
    },
    setAttribute(key, value) {
      obj[key] = value;
    },
  });

  obj.setAttribute('dataset', {});

  return obj;
}

const mockInsert = () => lineNumbers.push(mockLineNumberFactory());

function setContent(element, content) {
  /* eslint no-param-reassign:0 */
  // see comment in the source code being tested
  element.editor.value = content;
  element.lineCount = Math.max(1, content.split(/\n/).length);
  /* eslint no-param-reassign:1 */
}

describe('lineNumbersUpdate', () => {
  beforeEach(() => {
    lineNumbers = [];

    mockEditor = {
      editor: {
        style: {},
        value: '',
      },
      lineError: {},
      lineNumbers: {
        get offsetWidth() {
          return Math.max(1, lineNumbers.length) * 12; // totally made up
        },
        querySelectorAll() {
          return lineNumbers;
        },
      },
    };
  });

  it('should be a function', () => {
    expect(typeof lineNumbersUpdate).toBe('function');
  });

  it('should insert one lineNumber with no content', () => {
    setContent(mockEditor, '');

    lineNumbersUpdate(mockEditor, mockInsert);

    expect(mockEditor.editor.value).toBe('');
    expect(mockEditor.lineNumbers.querySelectorAll()).toMatchSnapshot('numbers');
    expect(mockEditor.editor.style).toMatchSnapshot('style');
  });

  it('should insert multiple lineNumbers with content', () => {
    const content = 'just\nsome\ncontent\nwith\nnewlines';

    setContent(mockEditor, content);

    lineNumbersUpdate(mockEditor, mockInsert);

    expect(mockEditor.editor.value).toBe(content);
    expect(mockEditor.lineNumbers.querySelectorAll()).toMatchSnapshot('numbers');
    expect(mockEditor.editor.style).toMatchSnapshot('style');
  });

  it('should show line errors', () => {
    const content = 'just\nsome\ncontent\nwith\nnewlines';

    setContent(mockEditor, content);
    mockEditor.lineError = {
      1: 'invalid for some reason',
      2: 'invalid because no numbers',
      3: 'invalid title',
      4: 'invalid line',
      5: 'invalid just because',
    };

    lineNumbersUpdate(mockEditor, mockInsert);

    expect(mockEditor.editor.value).toBe(content);
    expect(mockEditor.lineNumbers.querySelectorAll()).toMatchSnapshot('numbers');
    expect(mockEditor.editor.style).toMatchSnapshot('style');
  });

  it('should keep line errors after edit', () => {
    const content = 'just\nsome\ncontent\nwith\nnewlines';
    const appended = '\nsome\nnew\ncontent';

    setContent(mockEditor, content);
    mockEditor.lineError = {
      1: 'invalid for some reason',
      3: 'invalid title',
    };

    lineNumbersUpdate(mockEditor, mockInsert);

    mockEditor.editor.value += appended;

    expect(mockEditor.editor.value).toBe(content + appended);
    expect(mockEditor.lineNumbers.querySelectorAll()).toMatchSnapshot('numbers');
    expect(mockEditor.editor.style).toMatchSnapshot('style');
  });
});
