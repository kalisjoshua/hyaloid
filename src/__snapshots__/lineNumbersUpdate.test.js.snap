// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lineNumbersUpdate should insert multiple lineNumbers with content: numbers 1`] = `
Array [
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
]
`;

exports[`lineNumbersUpdate should insert multiple lineNumbers with content: style 1`] = `
Object {
  "marginLeft": "60px",
  "width": "calc(100% - 60px)",
}
`;

exports[`lineNumbersUpdate should insert one lineNumber with no content: numbers 1`] = `
Array [
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
]
`;

exports[`lineNumbersUpdate should insert one lineNumber with no content: style 1`] = `
Object {
  "marginLeft": "12px",
  "width": "calc(100% - 12px)",
}
`;

exports[`lineNumbersUpdate should keep line errors after edit: numbers 1`] = `
Array [
  Object {
    "dataset": Object {
      "hasError": true,
    },
    "title": "invalid for some reason",
  },
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
  Object {
    "dataset": Object {
      "hasError": true,
    },
    "title": "invalid title",
  },
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
  Object {
    "dataset": Object {
      "hasError": null,
    },
  },
]
`;

exports[`lineNumbersUpdate should keep line errors after edit: style 1`] = `
Object {
  "marginLeft": "60px",
  "width": "calc(100% - 60px)",
}
`;

exports[`lineNumbersUpdate should show line errors: numbers 1`] = `
Array [
  Object {
    "dataset": Object {
      "hasError": true,
    },
    "title": "invalid for some reason",
  },
  Object {
    "dataset": Object {
      "hasError": true,
    },
    "title": "invalid because no numbers",
  },
  Object {
    "dataset": Object {
      "hasError": true,
    },
    "title": "invalid title",
  },
  Object {
    "dataset": Object {
      "hasError": true,
    },
    "title": "invalid line",
  },
  Object {
    "dataset": Object {
      "hasError": true,
    },
    "title": "invalid just because",
  },
]
`;

exports[`lineNumbersUpdate should show line errors: style 1`] = `
Object {
  "marginLeft": "60px",
  "width": "calc(100% - 60px)",
}
`;
